doc/sphinx/development/tests.rst

```diff
@@ -1,41 +1,180 @@
+Writing tests
 =============
-Running Tests
-=============
 
-Using Docker
-============
+.. contents::
+
+When modifying code, adding features or fixing bugs, you're expected to write one or more tests to ensure that
+the feature your adding works correctly, or that the bug you've fixed doesn't come back.
+
+Integration tests modules live in the :file:`tests/functional/tests.d` folder.
+You may either add a new file to test your feature, or modify an existing file.
+
+These modules are shell scripts, and are sourced by the main integration test engine. Having a look at one of
+these modules will help you understand how they work, the :file:`tests/functional/tests.d/320-base.sh` is a good
+example you might want to look at.
+
+Example
+-------
+
+Here is a simple test taken from :file:`320-base.sh`:
+
+.. code-block:: none
+   :caption: a simple test
+
+   success   help2     $a0 --osh help
+   contain "OSH help"
+   json .error_code OK .command help .value null
+
+A complete reference of such commands can be found below, but let's explain this example in a few words:
+
+The command ``success`` implies that we're running a new test command, and that we expect it to work (we might
+also want to test invalid commands and ensure they fail as they should).
+The tester docker will connect to the target docker (that is running the bastion code) as a bastion user, and
+run the ``--osh help`` command there. This is expected to exit with a code indicating success (0),
+otherwise this test fails.
+
+The output of the command, once run on the bastion, should contain the text ``OSH help``, or the test will fail.
+
+In the JSON output (see :doc:`/using/api`) of this command, we expect to find the ``error_code`` field set to ``OK``,
+the ``command`` field set to ``help``, and the ``value`` field set to ``null``, or the test will fail.
+
+Running just this test will yield the following output:
+
+.. code-block:: none
+   :caption: a simple test output
+
+   00m04 [--] *** [0010/0021] 320-base::help2 (timeout --foreground 30 ssh -F /tmp/bastiontest.pgoA5h/ssh_config -i /tmp/bastiontest.pgoA5h/account0key1file user.5000@bastion_debian10_target -p 22 -- --json-greppable --osh help)
```

Comment:
Aaah I was wondering where you were seeing color codes, I guess you think `00m04` are color codes? They're not, it's a timer so that you can see in a glance how much time each tests takes!